### ISSUE TITLE
Move to concrete Send/Sync error types

### ISSUE THREAD

**Issue by alexcrichton** [rayon-rs/rayon#417]
One of the API guidelines is that errors should generally be `Send` and `Sync`, but it looks like `Box<Error>` is being returned in a few locations? Could a concrete error be returned instead?

**Comment by cuviper** [comment]
The places we currently use `Box<Error>` are in rayon-core, which we'd rather not bump semver.

Would it be a breaking change to make those `Box<Error + Send + Sync>`?  I think that may be OK -- at least this example does compile:

```rust
use std::error::Error;
fn foo() -> Box<Error + Send + Sync> {
    unimplemented!()
}
fn main() {
    let x: Box<Error> = foo();
}
```

**Comment by alexcrichton** [comment]
I think technically, yes, but practically, most likely no? There's a few options here though to do if a rayon-core bump is off the table. Most of them involve not directly reexporting rayon-core but rather wrapping it, but one example is that you could add a concrete, cloneable type in rayon-core. The rayon crate would then downcast to that type, clone it, and then return it (maybe wrapped again)

**Comment by nikomatsakis** [comment]
Hmm, the only place I see that we are using `Box<Error>` is the `initialize` and `ThreadPool::new` methods, right?

Maybe we can just deprecate those and create new names for them? Then we can use some concrete types or something? I'd like to be "failure compatible" I guess. Ultimately this doesn't seem like a super big deal though, those are pretty peripheral methods.

I do highly doubt that changing to `Box<Error + Send + Sync>` would break anyone. But I guess you never know.

**Comment by cuviper** [comment]
`initialize()`, `ThreadPool::new()`, and then `Configuration::build()` returns `ThreadPool::new()`.  One concrete type can cover all of these.

**Comment by nikomatsakis** [comment]
Decision:

`Configuration::build`, `ThreadPool::new`, and `initialize` would be deprecated. 

Add:

- `Configuration::init(self) -> ThreadPool`
- `Configuration::init_global(self)`

Alternative names:

- `Configuration::into_thread_pool(self) -> ThreadPool`
- `Configuration::into_global_thread_pool(self)`

**Comment by nikomatsakis** [comment]
Happy to bikeshed names further =)

**Comment by cuviper** [comment]
Those new methods should return `Result` with concrete error types.
(i.e. the whole point of this issue...)

**Comment by cuviper** [comment]
Fixed by #516.
